common/hashing_test.cpp

```diff
@@ -0,0 +1,677 @@
+                                       : "{0:x16}",
+                      static_cast<uint64_t>(value));
+}
+
+template <typename T, typename = std::enable_if_t<std::is_integral_v<T>>>
+auto operator<<(llvm::raw_ostream& os, HashedValue<T> hv)
+    -> llvm::raw_ostream& {
+  os << "hash " << hv.hash << " for value ";
+  PrintFullWidthHex(os, hv.v);
+  return os;
+}
+
+template <typename T, typename U,
+          typename = std::enable_if_t<std::is_integral_v<T>>,
+          typename = std::enable_if_t<std::is_integral_v<U>>>
+auto operator<<(llvm::raw_ostream& os, HashedValue<std::pair<T, U>> hv)
+    -> llvm::raw_ostream& {
+  os << "hash " << hv.hash << " for pair of ";
+  PrintFullWidthHex(os, hv.v.first);
+  os << " and ";
+  PrintFullWidthHex(os, hv.v.second);
+  return os;
+}
+
+struct Collisions {
+  int total;
+  int median;
+  int max;
+};
+
+// Analyzes a list of hashed values to find all of the hash codes which collide
+// within a specific bit-range.
+//
+// With `BitBegin=0` and `BitEnd=64`, this is equivalent to finding full
+// collisions. But when the begin and end of the bit range are narrower than the
+// 64-bits of the hash code, it allows this function to analyze a specific
+// window of bits within the 64-bit hash code to understand how many collisions
+// emerge purely within that bit range.
+//
+// With narrow ranges (we often look at the first N and last N bits for small
+// N), collisions are common and so this function summarizes this with the total
+// number of collisions and the median number of collisions for an input value.
+template <int BitBegin, int BitEnd, typename T>
+auto FindBitRangeCollisions(llvm::ArrayRef<HashedValue<T>> hashes)
+    -> Collisions {
+  static_assert(BitBegin < BitEnd);
+  constexpr int BitCount = BitEnd - BitBegin;
+  static_assert(BitCount <= 32);
+  constexpr int BitShift = BitBegin;
+  constexpr uint64_t BitMask = ((1ULL << BitCount) - 1) << BitShift;
+
+  // We collect counts of collisions in a vector. Initially, we just have a zero
+  // and all inputs map to that collision count. As we discover collisions,
+  // we'll create a dedicated counter for it and count how many inputs collide.
+  llvm::SmallVector<int> collision_counts;
+  collision_counts.push_back(0);
+  // The "map" for collision counts. Each input hashed value has a corresponding
+  // index stored here. That index is the index of the collision count in the
+  // container above. We resize this to fill it with zeros to start as the zero
+  // index above has a collision count of zero.
+  //
+  // The result of this is that the number of collisions for `hashes[i]` is
+  // `collision_counts[collision_map[i]]`.
+  llvm::SmallVector<int> collision_map;
+  collision_map.resize(hashes.size());
+
+  // First, we extract the bit subsequence we want to examine from each hash and
+  // store it with an index back into the hashed values (or the collision map).
+  llvm::SmallVector<std::pair<uint32_t, int>> bits_and_indices;
```

Comment:
Could we document the expected invariants here?
Initially `bits_and_indices[i].second == i`, but that isn't preserved. What we want to say is that `bits_and_indices[i].first` has hash bits from `hash[bits_and_indices[i].second]`.

It would be even better to use a struct instead of a pair and then document the meaning of the members of the struct.